pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+					}
+				}
+			}
+			<RelayEpoch<T>>::put(relay_epoch_index);
+			<InherentIncluded<T>>::put(());
+
+			Ok(Pays::No.into())
+		}
+	}
+
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
+		type Call = Call<T>;
+		type Error = InherentError;
+		const INHERENT_IDENTIFIER: InherentIdentifier = INHERENT_IDENTIFIER;
+
+		fn is_inherent_required(_: &InherentData) -> Result<Option<Self::Error>, Self::Error> {
+			// Return Ok(Some(_)) unconditionally because this inherent is required in every block
+			// If it is not found, throw a VrfInherentRequired error.
+			Ok(Some(InherentError::Other(
+				sp_runtime::RuntimeString::Borrowed(
+					"Inherent required to set babe randomness results",
+				),
+			)))
+		}
+
+		// The empty-payload inherent extrinsic.
+		fn create_inherent(_data: &InherentData) -> Option<Self::Call> {
+			Some(Call::set_babe_randomness_results {})
+		}
+
+		fn is_inherent(call: &Self::Call) -> bool {
+			matches!(call, Call::set_babe_randomness_results { .. })
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		// Set this block's randomness using the VRF output, verified by the VrfInput put in
+		// storage in the previous block's `on_initialize`
+		fn on_initialize(_now: BlockNumberFor<T>) -> Weight {
+			// Set and validate VRF output using input put in storage last block's `on_finalize`
+			vrf::set_output::<T>()
+		}
+		// Set next block's VRF input in storage
+		fn on_finalize(_now: BlockNumberFor<T>) {
+			// Panics if set_babe_randomness_results inherent was not included
+			assert!(
+				<InherentIncluded<T>>::take().is_some(),
+				"Mandatory randomness inherent not included; InherentIncluded storage item is empty"
+			);
+		}
+	}
+
+	// Utility function
+	impl<T: Config> Pallet<T> {
+		pub fn account_id() -> T::AccountId {
+			PALLET_ID.into_account_truncating()
+		}
+		pub fn total_locked() -> BalanceOf<T> {
+			// free balance is usable balance for the pallet account as it is not controlled
+			// by anyone so will never be locked
+			T::Currency::free_balance(&Self::account_id())
+		}
+		pub(crate) fn concat_and_hash(a: T::Hash, b: H256, index: u8) -> Vec<[u8; 32]> {
+			let mut output: Vec<[u8; 32]> = Vec::new();
+			for i in 0u8..index {
+				let mut s = Vec::new();
+				s.extend_from_slice(a.as_ref());
+				s.extend_from_slice(b.as_ref());
+				s.extend_from_slice(&[i]);
+				output.push(sp_io::hashing::blake2_256(&s))
+			}
+			output
+		}
+	}
+
+	// Public functions for precompile usage only
+	impl<T: Config> Pallet<T> {
+		pub fn request_randomness(
+			request: Request<BalanceOf<T>, RequestType<T>>,
+		) -> Result<RequestId, sp_runtime::DispatchError> {
+			let request = RequestState::new(request.into())?;
+			let (fee, contract_address, info) = (
+				request.request.fee,
+				request.request.contract_address,
+				request.request.info.clone(),
+			);
+			let total_to_reserve = T::Deposit::get().saturating_add(fee);
```

Comment:
Being paranoid again: I think erroring on overflow here is better. Otherwise it's possible to transfer less than the sum of `(deposit + fee)`, which would lead to an accounting discrepancy.

Reply:
It is saturating. The result does not overflow, it saturates to Balances::MAX which would lead to a fail at the transfer because they will not have that much balance.

Reply:
Yes, that's why I said I was being paranoid :) It's a defensive suggestion -- meant to reduce the probability of a bug. I agree that `transfer()` would fail if it saturates.

Also note that `fee` is arbitrary user input up to this point.